### ISSUE TITLE
Allow moving all files apart from assets out of web root

### ISSUE THREAD

**Issue by Findus23** [piwik/piwik#19505]
this is a bit of a continuation of #8120

At the moment, Matomo can only be used by settings the web root to the directory all Matomo files are in. This means additional web server configuration and precisely crafted rules are required so that security-relevant files (especially tmp/ and config/) are not publicly accessible. But as web server configurations vary widely and as the recent "private directories" system check shows, there are a lot of people for who these configurations don't work (most commonly because `AllowOverride` is disabled (#17819) and therefore the .htaccess files created by Matomo do nothing). And the system check that tries to access the URLs that should be private makes some people aware of this, but often confuses people (#18693) or creates new issues (#18182, #17589, #18967, #19149) either because the failing requests trigger things like fail2ban or because it exposes broken cacert-setups.
And that is completely ignoring the question of people who don't use apache and then have to find out themselves that they need to block some file requests or their Matomo setup is insecure.

To get to the point: I feel like offering a Matomo install which works similar to this (and how most modern PHP/laravel applications work) would solve a lot of complexity which gaining a lot of security:
- Matomo would have a public/ directory
- People could use Matomo by setting their webroot to this directory
- public/index.php would be a simple script that imports the main index.php and probably fixes some path
- and probably a config.ini.php option that would enable this mode (as otherwise "default" users would have duplicate URLs)
- JS/CSS assets would work out of the box as they use index.php
- other assets would need to be copied to a corresponding subpath in public/ either implicitly ("copy some directory in every plugin") or explicitly ("allow specifying a list of files that need to be public in the plugin description")
- While I don't think large changes are needed, these changes still feel breaking enough for me that this would need to be part of Matomo 5

Also this would be an optional feature as there are still people who don't know what a web root is (even though a lot of PHP FOSS applications don't offer any other way to be used).

In summary, I think this would make it a lot harder for people to accidentally shoot themselves into the foot in regard to security (it's not like I haven't noticed in the past that I had some important Matomo files accidentally public), but it also has the disadvantage of offering "one more way" to install Matomo.

**Comment by heurteph-ei** [comment]
@Findus23 , you write:
> - JS/CSS assets would work out of the box as they use index.php

Do you mean Matomo tracker JS file and MTM containers?
If yes, this will then prevent the easy publication of MTM containers through CDN for example...

**Comment by Findus23** [comment]
> Do you mean Matomo tracker JS file and MTM containers?
> If yes, this will then prevent the easy publication of MTM containers through CDN for example...

When I wrote this, I didn't think of those. I just mean the JS and the CSS that is used by the Matomo Backend.
It's true that matomo.js (and all variants of those) and the MTM containers would need to be inside the public root. But that should not be more complex to solve than the rest of this issue.

**Comment by MatomoForumNotifications** [comment]
This issue has been mentioned on **Matomo forums**. There might be relevant details there:

https://forum.matomo.org/t/how-to-harden-a-matomo-installation/49765/6

**Comment by MatomoForumNotifications** [comment]
This issue has been mentioned on **Matomo forums**. There might be relevant details there:

https://forum.matomo.org/t/impossible-to-install-the-wordpress-plugin-on-elementor/49767/4

**Comment by MatomoForumNotifications** [comment]
This issue has been mentioned on **Matomo forums**. There might be relevant details there:

https://forum.matomo.org/t/required-private-directories-recommended-private-directories/50193/2

**Comment by randy-innocraft** [comment]
Hi @Findus23 ,
Thank you for bringing this to our attention and for your valuable input.
Your suggestion seems like a valuable enhancement to our product.
We will forward this to our Product team for review and future consideration.
If you have any additional details or questions, please feel free to share them here.

**Comment by joelpittet** [comment]
Moving the `vendor` directory out of the web root is good practice too, glad this has already brought up. 

You could do it in a way where you could do both, [roundcube](https://github.com/roundcube/roundcubemail/) does this. So you can do the sub directory for security and keep the current setup as to not disrupt the existing deployments.

https://github.com/roundcube/roundcubemail/blob/master/public_html/index.php
https://github.com/roundcube/roundcubemail/blob/master/index.php

I prefer calling it `public` like the OP suggested FWIW.

**Comment by ksubileau** [comment]
I would like to add my support to this request.

It is now common practice, and considered a good security practice, to limit publicly accessible resources by placing them in a specific directory, and to move all PHP source code outside of this public path — in particular configuration files that may contain secrets, such as config.ini.php.

Most popular PHP frameworks, as well as frameworks in other languages, have adopted this structure, including Symfony, Laravel, and others.

Relying on htaccess files — which only work with Apache — or on a documented Nginx configuration that is merely recommended, seems to me like an outdated approach that is insufficient and no longer aligned with current best practices. This seems quite dangerous and doesn't inspire confidence in the product's security compared to other available solutions.

In my case, which is far from unusual in modern architectures, I host Matomo on a Kubernetes cluster with Traefik as the ingress controller. To minimize resource usage, I chose FrankenPHP (which is based on Caddy) to serve Matomo behind Traefik, which is much lighter than Apache or Nginx. In this setup, neither the documented configuration nor the .htaccess files apply. Sensitive files end up exposed to the internet by default. I had to search for and implement my own solution to block access to these sensitive paths, and I remain relatively unconfident about the overall security, knowing that even the slightest configuration mistake in this single layer of protection could expose critical data.

Therefore, I believe that revising Matomo’s file structure is essential — at the latest for the next major version — to bring the product up to current standards. Otherwise, this could clearly be a reason for me to stop using Matomo in favor of competitors that adopt a more modern, security-by-design approach.

**Comment by tassoman** [comment]
I guess the best practice for orchestration is to share the entire config file as a _Secret_.  

The biggest part of hosting setups are marketed without a console or VPS, or containers.  
Today this setup let's onboard the most of the people by running a cheap hosting, self hosting (in a nas).  
High profile users are still able to customise their needs (as you noted). 
I guess, the main goal of the project is to embrace most of the people letting them unchain/unlock from the big companies of analytics.
By the way, thanks for sharing your thoughts, I agree can be a kind of difficult configuring an orchestrated version Matomo. 
But I guess it's just the minority part of users who configure it in a _complex_ environment.

**Comment by ksubileau** [comment]
> I guess the best practice for orchestration is to share the entire config file as a Secret.

Yes, indeed, the configuration on Kubernetes is usually provided through a ConfigMap or a Secret. 
On that point as well, Matomo is somewhat behind in terms of containerization capabilities compared to competitors like Unami or Plausible (the provided Docker image is oddly built and incomplete in terms of configuration options; the initial installation requires an “ExtraTools” plugin, etc.; see https://github.com/matomo-org/matomo/issues/10257). But that’s another topic.

> The biggest part of hosting setups are marketed without a console or VPS, or containers.
Today this setup let's onboard the most of the people by running a cheap hosting, self hosting (in a nas).

I myself use low-cost hosting — both shared hosting without console access and a low-tier VPS with K3S, a lightweight Kubernetes distribution. Both allow me to change the web root if necessary.

I believe that most hosting providers, even low-cost ones, allow at least changing the public root directory of the hosted service. And in the worst case, I think it's still possible to offer a fallback solution for those who can’t do that (for example, an index.php in the parent folder of the public directory that loads the index.php from the public directory, as said before).
I think that setting aside a major security best practice for the benefit of all users for the sake of a minority who could not implement it would be a strange choice.
